src/libstore/local-store.cc

```diff
@@ -1343,9 +1354,15 @@ void LocalStore::addToStore(const ValidPathInfo & info, Source & source,
             optimisePath(realPath, repair); // FIXME: combine with hashPath()
 
             registerValidPath(info);
+        } else {
+            /* Discard the NAR from the source if it has been added already. */
+            discardNar(source, info);
         }
 
         outputLock.setDeletion(true);
+    } else {
+        /* Discard the NAR from the source if it has been added already. */
+        discardNar(source, info);
```

Comment:
Instead of having two calls to `discardNar()`, it might be nicer to do something like:
```
bool narRead = false;
Finally cleanup = [&]() {
    if (!narRead) {
        ParseSink sink; /* null sink; just parse the NAR */
        parseDump(sink, source);
    }
};
```
and then set `narRead = true` before the call to `restorePath()`. That way it will keep working if we ever add other return paths out of this function.

Reply:
Yeah, that does sound much nicer